### PR TITLE
[CHANGED] StartAtSequence and StartAtTimeDelta no longer fail if out …

### LINKED ISSUE #7: Start time/sequence validation should be "inclusive" vs. "exact"
_From @mcqueary on March 15, 2016 19:14_

See [here](https://github.com/nats-io/stan/blob/master/server.go#L1061-L1070) and [here](https://github.com/nats-io/stan/blob/master/server.go#L1226-L1235) in `server.go`.

As a user, if my `SubscriptionRequest` specifies a time 100 seconds in the past, but the earliest stored message was 80 seconds in the past, I get `stan: invalid start time`. IMO that's broken, conceptually. 

Think of how you'd use a "since" date when qualifying Google search results (i.e. show me hits since this date). If I entered 1/1/2016, but the first result was from 1/3/2016, I'd expect to get everything starting from 1/3/2016, and not an error indicating that my "since" date was wrong. The same applies here: if my start time is before the earliest message, just start with the earliest message!

It's tempting to say "well, just use StartPosition.First instead", but I as a user actually require data since exactly (at most) 100 seconds ago, and I have no idea when "First" will start (3 days, 2 hours, or 5 seconds). 

I propose that if start time (or sequence) is earlier than the first message, that we start with the first message, and likewise that if the start time (or sequence) is later than the last message, we set the start time/sequence to 1ns or 1 seqno after the last message (i.e. default to `StartPosition.NewOnly`).

_Copied from original issue: nats-io/stan#24_

### PR DESCRIPTION
**Pull request by kozlovic**
…of range

StartAtSequence will now start delivery at first message if given
sequence is smaller than first available, and equivalent to "NewOnly"
if given sequence is greater than current last sequence.

Similar behavior for StartAtTimeDelta.

Resolves #7

### CODE REVIEW

**Comment by sfrooster** [comment]
@kozlovic 
When is this scheduled to be merged? :-)

**Comment by kozlovic** [comment]
@aricart @ColinSullivan1 @mcqueary If you would like to have a look before merging?

File: server/server.go [hunk 1]
```diff
@@ -2632,14 +2587,29 @@ func (s *StanServer) setSubStartSequence(cs *stores.ChannelStore, sub *subState,
...
+			// so this would translate to "new only" semantic.
 			lastSent = seq - 1
 		}
 		Debugf("STAN: [Client:%s] Sending from time, subject=%s time=%d seq=%d",
 			sub.ClientID, sub.subject, startTime, lastSent)
 	case pb.StartPosition_SequenceStart:
-		if sr.StartSequence > 0 {
+		// If there is no message, firstSeq and lastSeq will be equal to 0.
+		firstSeq, lastSeq := cs.Msgs.FirstAndLastSequence()
+		// StarSequence is an uint64, so can't be lower than 0.
```
Comment (ColinSullivan1) [inline server/server.go]: Typo in comment.

File: server/server.go [hunk 2]
```diff
@@ -2632,14 +2587,29 @@ func (s *StanServer) setSubStartSequence(cs *stores.ChannelStore, sub *subState,
...
+		if sr.StartSequence < firstSeq {
```
Comment (ColinSullivan1) [inline server/server.go]: I'd suggest a debug message.  Perhaps even an option to dictate behavior here?  The concern lies in the case where a log rollover occurred - some users would want to know they are not receiving all messages in their stream.

File: server/server_test.go [hunk 3]
```diff
@@ -2090,42 +2090,81 @@ func TestStartPositionFirstSequence(t *testing.T) {
...
-	// As of now, since there is no message, the call will fail.
-	sub, err := sc.Subscribe("foo", cb, stan.StartAtSequence(0))
-	if err == nil {
-		sub.Unsubscribe()
-		t.Fatal("Expected error on subscribe, got none")
+	rch := make(chan bool)
+	// Create subscriber with sequence below firstSeq, it should not fail
+	// and receive message with sequence == firstSeq
+	sub, err := sc.Subscribe("foo", func(m *stan.Msg) {
+		if m.Sequence == firstSeq {
```
Comment (ColinSullivan1) [inline server/server_test.go]: I believe you should also check that the first message received is the firstSeq.  Grated, other tests ensure the store has rolled of messages, but it'd be nice.

File: server/server_test.go [hunk 4]
```diff
@@ -2090,42 +2090,81 @@ func TestStartPositionFirstSequence(t *testing.T) {
...
+	if err := Wait(rch); err != nil {
+		t.Fatal("Did not get our message")
+	}
+	sub.Unsubscribe()
 
-	// Send a message now.
+	// Create subscriber with sequence higher than lastSeq, it should not
+	// fail, but not receive until we send a new message.
+	sub, err = sc.Subscribe("foo", func(m *stan.Msg) {
+		if m.Sequence == lastSeq+1 {
```
Comment (ColinSullivan1) [inline server/server_test.go]: You'll want to also check to see it's the first message received by the subscriber.

File: server/server_test.go [hunk 5]
```diff
@@ -2137,39 +2176,68 @@ func TestStartPositionTimeDelta(t *testing.T) {
...
 	if err := sc.Publish("foo", []byte("msg2")); err != nil {
 		t.Fatalf("Unexpected error on publish: %v", err)
 	}
 
-	// Start a subscriber with "TimeDelta" as start position.
-	sub, err := sc.Subscribe("foo", cb, stan.StartAtTimeDelta(1*time.Second))
+	// Create subscriber with TimeDelta in the past, should
+	// not fail and get first message.
+	sub, err := sc.Subscribe("foo", func(m *stan.Msg) {
+		if m.Sequence == 1 {
```
Comment (ColinSullivan1) [inline server/server_test.go]: ...check that it is the first message received (see above).

File: server/server.go (hunk 1 above)
Comment (kozlovic) [inline server/server.go]: Done

File: server/server_test.go (hunk 3 above)
Comment (kozlovic) [inline server/server_test.go]: Done

File: server/server_test.go (hunk 4 above)
Comment (kozlovic) [inline server/server_test.go]: Done

File: server/server_test.go (hunk 5 above)
Comment (kozlovic) [inline server/server_test.go]: Done

**Comment by ColinSullivan1** [comment]
LGTM!

**Pull request merged**